csharp/src/Ice/OutputStream.cs

```diff
@@ -169,232 +111,171 @@ public byte[] Finished()
-            return _buf;
-        }
+            Encoding = encoding;
+            if (format.HasValue)
+            {
+                _format = format.Value;
+            }
 
-        /// <summary>
-        /// Marks the start of a class instance.
-        /// </summary>
-        /// <param name="data">Preserved slices for this instance, or null.</param>
-        public void StartClass(SlicedData? data)
-        {
-            Debug.Assert(_encapsStack != null && _encapsStack.encoder != null);
-            _encapsStack.encoder.StartInstance(SliceType.ClassSlice, data);
+            WriteEncapsulationHeader(0, Encoding);
         }
 
         /// <summary>
-        /// Marks the end of a class instance.
+        /// Ends the previous main encapsulation.
         /// </summary>
-        public void EndClass()
+        public void EndEncapsulation()
         {
-            Debug.Assert(_encapsStack != null && _encapsStack.encoder != null);
-            _encapsStack.encoder.EndInstance();
-        }
+            Debug.Assert(_mainEncaps.HasValue && _endpointEncaps == null);
 
-        /// <summary>
-        /// Marks the start of a user exception.
-        /// </summary>
-        /// <param name="data">Preserved slices for this exception, or null.</param>
-        public void StartException(SlicedData? data)
-        {
-            Debug.Assert(_encapsStack != null && _encapsStack.encoder != null);
-            _encapsStack.encoder.StartInstance(SliceType.ExceptionSlice, data);
-        }
+            // Size includes size and version.
+            int startPos = _mainEncaps.Value.StartPos;
+            int sz = _buf.size() - startPos;
+            _buf.b.putInt(startPos, sz);
 
-        /// <summary>
-        /// Marks the end of a user exception.
-        /// </summary>
-        public void EndException()
-        {
-            Debug.Assert(_encapsStack != null && _encapsStack.encoder != null);
-            _encapsStack.encoder.EndInstance();
+            Encoding = _mainEncaps.Value.OldEncoding;
+            _format = _mainEncaps.Value.OldFormat;
+            ResetEncapsulation();
         }
 
         /// <summary>
-        /// Writes the start of an encapsulation to the stream.
+        /// Start writing a slice of a class or exception instance.
+        /// This is an Ice-internal method marked public because it's called by the generated code.
         /// </summary>
-        public void StartEncapsulation()
+        /// <param name="typeId">The type ID of this slice.</param>
+        /// <param name="firstSlice">True when writing the first (most derived) slice of an instance.</param>
+        /// <param name="slicedData">Preserved sliced-off slices, if any. Can only be provided when firstSlice is true.
+        /// </param>
+        /// <param name="compactId">The compact type ID of this slice, if specified."</param>
+        public void IceStartSlice(string typeId, bool firstSlice, SlicedData? slicedData = null, int? compactId = null)
```

Comment:
Should we mark Ice prefixed methods with `[EditorBrowsable(EditorBrowsableState.Never)]` like in the generated code?

Reply:
Yes, fixed.